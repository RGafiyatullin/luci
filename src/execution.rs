use std::{
    collections::{BTreeSet, HashMap},
    sync::Arc,
    time::Duration,
};

use serde_json::Value;
use slotmap::{new_key_type, SlotMap};

use crate::{
    names::{ActorName, EventName},
    scenario::{Msg, RequiredToBe},
};

mod build;
mod report;
mod runner;

pub use build::BuildError;
pub use report::Report;
pub use runner::RunError;
pub use runner::Runner;

new_key_type! {
    pub struct KeyBind;
    pub struct KeySend;
    pub struct KeyRecv;
    pub struct KeyRespond;
    pub struct KeyDelay;
}

/// A key corresponding to some event during test execution.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum EventKey {
    Bind(KeyBind),
    Send(KeySend),
    Recv(KeyRecv),
    Respond(KeyRespond),
    Delay(KeyDelay),
}

#[derive(Debug)]
pub struct Executable {
    events: Events,
}

#[derive(Debug, Default)]
struct Events {
    priority: HashMap<EventKey, usize>,
    required: HashMap<EventKey, RequiredToBe>,
    names: HashMap<EventKey, EventName>,

    bind: SlotMap<KeyBind, EventBind>,
    send: SlotMap<KeySend, EventSend>,
    recv: SlotMap<KeyRecv, EventRecv>,
    respond: SlotMap<KeyRespond, EventRespond>,
    delay: SlotMap<KeyDelay, EventDelay>,

    entry_points: BTreeSet<EventKey>,

    key_unblocks_values: HashMap<EventKey, BTreeSet<EventKey>>,
}

#[derive(Debug)]
struct EventSend {
    from: ActorName,
    to: Option<ActorName>,
    fqn: Arc<str>,
    payload: Msg,
}

#[derive(Debug)]
struct EventRecv {
    from: Option<ActorName>,
    to: Option<ActorName>,
    fqn: Arc<str>,
    payload: Msg,
}

#[derive(Debug)]
struct EventRespond {
    respond_to: KeyRecv,
    request_type: Arc<str>,
    respond_from: Option<ActorName>,
    payload: Msg,
}

#[derive(Debug)]
struct EventBind {
    dst: Value,
    src: Msg,
}

#[derive(Debug)]
struct EventDelay {
    delay_for: Duration,
    delay_step: Duration,
}
