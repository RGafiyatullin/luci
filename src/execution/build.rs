//! This module is responsible for building an [`Executable`] from [`Sources`].
//!

use std::{
    collections::{BTreeSet, HashMap, HashSet},
    sync::Arc,
};

use serde_json::json;
use slotmap::SlotMap;
use tracing::{debug, trace};

use crate::{
    execution::{
        BindScope, EventBind, EventKey, KeyBind, KeyDelay, KeyRecv, KeyRespond, KeyScope, KeySend,
        KeySource, ScopeInfo, Sources,
    },
    marshalling,
    names::SubroutineName,
    scenario::{
        DefEventBind, DefEventDelay, DefEventRecv, DefEventRespond, DefEventSend, Msg, RequiredToBe,
    },
};
use crate::{
    execution::{EventDelay, EventRecv, EventRespond, EventSend, Events, Executable},
    marshalling::MarshallingRegistry,
    names::{ActorName, EventName, MessageName},
    scenario::{DefEvent, DefEventKind, DefTypeAlias},
};

#[derive(Debug, thiserror::Error)]
pub enum BuildError {
    #[error("unknown event: {}", _0)]
    UnknownEvent(EventName),

    #[error("duplicate event: {}", _0)]
    DuplicateEventName(EventName),

    #[error("not a request: {}", _0)]
    NotARequest(EventName),

    #[error("unknown actor: {}", _0)]
    UnknownActor(ActorName),

    #[error("unknown subroutine: {}", _0)]
    UnknownSubroutine(SubroutineName),

    #[error("unknown FQN: {}", _0)]
    UnknownFqn(String),

    #[error("unknown alias: {}", _0)]
    UnknownAlias(MessageName),

    #[error("duplicate alias: {}", _0)]
    DuplicateAlias(MessageName),

    #[error("duplicate actor name: {}", _0)]
    DuplicateActorName(ActorName),

    #[error("invalid data: {}", _0)]
    InvalidData(marshalling::AnError),
}

impl Executable {
    pub fn build<'a>(
        marshalling: MarshallingRegistry,
        sources: &'a Sources,
        main: KeySource,
    ) -> Result<Self, BuildError> {
        debug!("building...");

        let mut builder: Builder = Default::default();

        let SubgraphAdded {
            scope_key,
            entry_points,
            require: required,
        } = builder.add_subgraph(&marshalling, sources, main, None)?;
        let Builder {
            scopes,
            event_names,
            definition_order,
            events_delay,
            events_bind,
            events_recv,
            events_send,
            events_respond,
            key_unblocks_values,
        } = builder;

        let priority = definition_order
            .into_iter()
            .enumerate()
            .map(|(p, k)| (k, p))
            .collect();

        let events = Events {
            priority,
            required,
            names: event_names,
            bind: events_bind,
            send: events_send,
            recv: events_recv,
            respond: events_respond,
            delay: events_delay,
            entry_points,
            key_unblocks_values,
        };

        Ok(Executable {
            marshalling,
            events,
            root_scope_key: scope_key,
            scopes,
        })
    }
}

fn type_aliases<'a>(
    marshalling: &MarshallingRegistry,
    imports: impl IntoIterator<Item = &'a DefTypeAlias>,
) -> Result<HashMap<MessageName, Arc<str>>, BuildError> {
    use std::collections::hash_map::Entry::Vacant;
    let mut aliases = HashMap::new();
    for import in imports {
        let Vacant(entry) = aliases.entry(import.type_alias.to_owned()) else {
            return Err(BuildError::DuplicateAlias(import.type_alias.clone()));
        };
        let _marshaller = marshalling
            .resolve(&import.type_name)
            .ok_or(BuildError::UnknownFqn(import.type_name.to_owned()))?;

        entry.insert(import.type_name.as_str().into());
    }

    Ok(aliases)
}

fn validate_actor_names<'a>(
    actor_names: impl IntoIterator<Item = &'a ActorName>,
) -> Result<HashSet<ActorName>, BuildError> {
    let mut out = HashSet::new();

    for name in actor_names {
        if !out.insert(name.clone()) {
            return Err(BuildError::DuplicateActorName(name.clone()));
        }
    }

    Ok(out)
}

fn resolve_event_ids<'a>(
    idx_keys: &'a HashMap<&'a EventName, EventKey>,
    names: &'a [EventName],
) -> impl Iterator<Item = Result<EventKey, BuildError>> + 'a {
    names.into_iter().map(move |name: &EventName| {
        idx_keys
            .get(name)
            .copied()
            .ok_or(BuildError::UnknownEvent(name.clone()))
    })
}

#[derive(Debug, Default)]
struct Builder {
    scopes: SlotMap<KeyScope, ScopeInfo>,

    event_names: HashMap<EventKey, (KeyScope, EventName)>,

    definition_order: Vec<EventKey>,

    events_delay: SlotMap<KeyDelay, EventDelay>,
    events_bind: SlotMap<KeyBind, EventBind>,
    events_recv: SlotMap<KeyRecv, EventRecv>,
    events_send: SlotMap<KeySend, EventSend>,
    events_respond: SlotMap<KeyRespond, EventRespond>,

    key_unblocks_values: HashMap<EventKey, BTreeSet<EventKey>>,
}

#[derive(Debug)]
struct SubgraphAdded {
    scope_key: KeyScope,
    entry_points: BTreeSet<EventKey>,
    require: HashMap<EventKey, RequiredToBe>,
}

impl Builder {
    fn add_subgraph(
        &mut self,
        marshalling: &MarshallingRegistry,
        sources: &Sources,
        source_key: KeySource,
        invoked_as: Option<(KeyScope, EventName, SubroutineName)>,
    ) -> Result<SubgraphAdded, BuildError> {
        let this_source = &sources[source_key];

        debug!("storing type-aliases...");
        let type_aliases = type_aliases(&marshalling, &this_source.scenario.types)?;
        for (a, fqn) in &type_aliases {
            trace!("- {:?} -> {:?}", a, fqn);
        }

        debug!("checking actor-names...");
        let actors = validate_actor_names(&this_source.scenario.cast)?;
        for actor_name in &actors {
            trace!("- {:?}", actor_name);
        }

        let this_scope_key = self.scopes.insert(ScopeInfo {
            source_key,
            invoked_as,
        });

        let mut this_scope_name_to_key = HashMap::new();
        let mut this_scope_entry_points = BTreeSet::new();
        let mut this_scope_requires = HashMap::new();

        for DefEvent {
            id: this_name,
            require: this_event_required_to_be,
            prerequisites,
            kind,
            ..
        } in this_source.scenario.events.iter()
        {
            let prerequisites = resolve_event_ids(&mut this_scope_name_to_key, &prerequisites)
                .collect::<Result<Vec<_>, _>>()?;

            let (head_key, tail_key) = match kind {
                DefEventKind::Call(def_call) => {
                    let sub_source_key = this_source
                        .subs
                        .get(&def_call.subroutine_name)
                        .copied()
                        .ok_or_else(|| {
                        BuildError::UnknownSubroutine(def_call.subroutine_name.clone())
                    })?;
                    let SubgraphAdded {
                        scope_key: sub_scope_key,
                        entry_points: sub_entry_points,
                        require: sub_required_to_be,
                    } = self.add_subgraph(
                        marshalling,
                        sources,
                        sub_source_key,
                        Some((
                            this_scope_key,
                            this_name.clone(),
                            def_call.subroutine_name.clone(),
                        )),
                    )?;

                    // create two bind nodes:
                    // - one for input (bind from `scope_key` to `sub_scope_key`, choose the nodes using `entrypoints`)
                    // - one for output (bind from `sub_scope_key` to `scope_key`, choose the nodes using `required`)
                    //
                    // the latter bind will be referred to by `this_key`, so that it can be depended on
                    // (the events that want to happen after this call — should take place after the output-bind).

                    let event_bind_in = {
                        let (dst, src) = if let Some(def_bind_in) = def_call.input.as_ref() {
                            (def_bind_in.dst.clone(), Msg::Bind(def_bind_in.src.clone()))
                        } else {
                            (json!(null), Msg::Literal(json!(null)))
                        };
                        EventBind {
                            dst,
                            src,
                            scope: BindScope::Two {
                                src: this_scope_key,
                                dst: sub_scope_key,
                                actors: def_call.cast.clone().unwrap_or_default(),
                            },
                        }
                    };
                    let bind_in = self.events_bind.insert(event_bind_in);
                    let ek_bind_in = EventKey::Bind(bind_in);

                    for sub_entry_point in sub_entry_points {
                        let hasnt_been_added_before = self
                            .key_unblocks_values
                            .entry(ek_bind_in)
                            .or_default()
                            .insert(sub_entry_point);
                        assert!(hasnt_been_added_before);
                    }

                    let event_bind_out = {
                        let (dst, src) = if let Some(def_bind_out) = def_call.output.as_ref() {
                            (
                                def_bind_out.dst.clone(),
                                Msg::Bind(def_bind_out.src.clone()),
                            )
                        } else {
                            (json!(null), Msg::Literal(json!(null)))
                        };
                        EventBind {
                            dst,
                            src,
                            scope: BindScope::Two {
                                src: sub_scope_key,
                                dst: this_scope_key,
                                actors: def_call
                                    .cast
                                    .clone()
                                    .unwrap_or_default()
                                    .into_iter()
                                    .map(|(l, r)| (r, l))
                                    .collect(),
                            },
                        }
                    };
                    let bind_out = self.events_bind.insert(event_bind_out);
                    let ek_bind_out = EventKey::Bind(bind_out);

                    for (sub_key, requirement) in sub_required_to_be {
                        if matches!(requirement, RequiredToBe::Reached) {
                            let hasnt_been_added_before = self
                                .key_unblocks_values
                                .entry(sub_key)
                                .or_default()
                                .insert(ek_bind_out);
                            assert!(hasnt_been_added_before);
                        }
                    }

                    (ek_bind_in, ek_bind_out)
                }
                DefEventKind::Delay(def_delay) => {
                    let DefEventDelay {
                        delay_for,
                        delay_step,
                        no_extra: _,
                    } = def_delay;
                    let delay_for = *delay_for;
                    let delay_step = *delay_step;

                    let key = self.events_delay.insert(EventDelay {
                        delay_for,
                        delay_step,
                    });
                    let ek_delay = EventKey::Delay(key);
                    (ek_delay, ek_delay)
                }
                DefEventKind::Bind(def_bind) => {
                    let DefEventBind {
                        dst,
                        src,
                        no_extra: _,
                    } = def_bind;
                    let dst = dst.clone();
                    let src = src.clone();
                    let key = self.events_bind.insert(EventBind {
                        dst,
                        src,
                        scope: BindScope::Same(this_scope_key),
                    });

                    let ek_bind = EventKey::Bind(key);
                    (ek_bind, ek_bind)
                }
                DefEventKind::Recv(def_recv) => {
                    let DefEventRecv {
                        message_type,
                        message_data,
                        from,
                        to,
                        no_extra: _,
                    } = def_recv;

                    let type_fqn = type_aliases
                        .get(message_type)
                        .cloned()
                        .ok_or(BuildError::UnknownAlias(message_type.clone()))?;

                    for actor_name in to.as_ref().into_iter().chain(from) {
                        if !actors.contains(actor_name) {
                            return Err(BuildError::UnknownActor(actor_name.clone()));
                        }
                    }

                    let key = self.events_recv.insert(EventRecv {
                        from: from.clone(),
                        to: to.clone(),
                        fqn: type_fqn,
                        payload: message_data.clone(),
                        scope_key: this_scope_key,
                    });
                    let ek_recv = EventKey::Recv(key);
                    (ek_recv, ek_recv)
                }
                DefEventKind::Respond(def_respond) => {
                    let DefEventRespond {
                        from,
                        to_request: to,
                        data,
                        no_extra: _,
                    } = def_respond;

                    let causing_event_key = this_scope_name_to_key
                        .get(&to)
                        .ok_or(BuildError::UnknownEvent(to.clone()))?;
                    let EventKey::Recv(recv_key) = causing_event_key else {
                        return Err(BuildError::NotARequest(to.clone()));
                    };
                    let request_fqn = self.events_recv.get(*recv_key)
                        .expect("we do not delete items from `recv`; neither we store keys that are unrelated to our collections")
                        .fqn.clone();

                    if let Some(bad_actor) = from.as_ref().filter(|a| !actors.contains(a)) {
                        return Err(BuildError::UnknownActor(bad_actor.clone()));
                    }

                    if marshalling
                        .resolve(&request_fqn)
                        .is_none_or(|m| m.response().is_none())
                    {
                        return Err(BuildError::NotARequest(to.clone()));
                    }

                    let key = self.events_respond.insert(EventRespond {
                        respond_to: *recv_key,
                        request_type: request_fqn,
                        respond_from: from.clone(),
                        payload: data.clone(),
                        scope_key: this_scope_key,
                    });
                    let ek_respond = EventKey::Respond(key);
                    (ek_respond, ek_respond)
                }
                DefEventKind::Send(def_send) => {
                    let DefEventSend {
                        from,
                        to,
                        message_type,
                        message_data,
                        no_extra: _,
                    } = def_send;

                    let type_fqn = type_aliases
                        .get(message_type)
                        .cloned()
                        .ok_or(BuildError::UnknownAlias(message_type.clone()))?;

                    for actor_name in to.as_ref().into_iter().chain([from]) {
                        if !actors.contains(&actor_name) {
                            return Err(BuildError::UnknownActor(actor_name.clone()));
                        }
                    }

                    let key = self.events_send.insert(EventSend {
                        from: from.clone(),
                        to: to.clone(),
                        fqn: type_fqn,
                        payload: message_data.clone(),
                        scope_key: this_scope_key,
                    });
                    let ek_send = EventKey::Send(key);
                    (ek_send, ek_send)
                }
            };

            if let Some(r) = this_event_required_to_be {
                this_scope_requires.insert(tail_key, *r);
            }

            if prerequisites.is_empty() {
                let should_be_a_new_element = this_scope_entry_points.insert(head_key);
                assert!(
                    should_be_a_new_element,
                    "non unique entry point? {:?}",
                    head_key
                );
            }
            for prerequisite in &prerequisites {
                let should_be_a_new_element = self
                    .key_unblocks_values
                    .entry(*prerequisite)
                    .or_default()
                    .insert(head_key);

                assert!(
                    should_be_a_new_element,
                    "duplicate  relation: {:?} unblocks {:?}",
                    *prerequisite, head_key
                );
            }

            trace!("  done: {:?} -> {:?}-{:?}", this_name, head_key, tail_key);

            if this_scope_name_to_key.insert(this_name, tail_key).is_some() {
                return Err(BuildError::DuplicateEventName(this_name.clone()));
            }
            self.definition_order.push(head_key);
            self.definition_order.push(tail_key);
        }

        for (name, key) in this_scope_name_to_key {
            let should_be_none = self.event_names.insert(key, (this_scope_key, name.clone()));
            assert!(should_be_none.is_none());
        }

        Ok(SubgraphAdded {
            scope_key: this_scope_key,
            entry_points: this_scope_entry_points,
            require: this_scope_requires,
        })
    }
}
