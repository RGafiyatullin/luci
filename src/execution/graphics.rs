use std::collections::HashSet;

use crate::graphics::{OutputFormat, RenderGraph};

use super::{
    EventBind, EventDelay, EventKey, EventRecv, EventRespond, EventSend, Events, Executable,
};

impl RenderGraph for Executable {
    const OUTPUT: OutputFormat = OutputFormat::Graphviz;

    fn render(&self) -> String {
        let mut acc = String::new();
        acc.push_str("digraph test { rankdir=LR layout=dot\n");

        self.events
            .entry_points
            .iter()
            .chain(self.events.key_unblocks_values.values().flatten())
            .cloned()
            .collect::<HashSet<EventKey>>() // deduplicate
            .iter()
            .for_each(|key| {
                self.events.draw_node(&mut acc, &key);
            });

        for (parent, children) in &self.events.key_unblocks_values {
            for child in children {
                acc.push_str(&format!(r#"  "{:?}" -> "{:?}""#, parent, child));
            }
        }

        acc.push_str("}\n");
        acc
    }
}

impl Events {
    fn draw_node(&self, acc: &mut String, key: &EventKey) {
        let node = match key {
            EventKey::Delay(key_delay) => {
                let delay = self.delay.get(*key_delay).unwrap();
                delay.draw(*key)
            }
            EventKey::Bind(key_bind) => {
                let bind = self.bind.get(*key_bind).unwrap();
                bind.draw(*key)
            }
            EventKey::Recv(key_recv) => {
                let recv = self.recv.get(*key_recv).unwrap();
                recv.draw(*key)
            }
            EventKey::Send(key_send) => {
                let send = self.send.get(*key_send).unwrap();
                send.draw(*key)
            }
            EventKey::Respond(key_respond) => {
                let respond = self.respond.get(*key_respond).unwrap();
                respond.draw(*key)
            }
        };
        acc.push_str(&format!("{}", &node));
        acc.push('\n');
    }
}

pub trait DrawDot {
    fn draw(&self, key: EventKey) -> String;
}

impl DrawDot for EventDelay {
    fn draw(&self, key: EventKey) -> String {
        format!(
            r#""{:?}" [label="delay {:?} by {:?}"]"#,
            key, self.delay_for, self.delay_step
        )
    }
}

impl DrawDot for EventBind {
    fn draw(&self, key: EventKey) -> String {
        let src = serde_yaml::to_string(&self.src).unwrap();
        let dst = serde_yaml::to_string(&self.dst).unwrap();
        format!(
            r#""{:?}" [label="bind\nsrc: \n{}\ndst: \n{}"]"#,
            key, src, dst
        )
    }
}

impl DrawDot for EventRecv {
    fn draw(&self, key: EventKey) -> String {
        let data = serde_yaml::to_string(&self.payload).unwrap();
        format!(
            r#""{:?}" [label="recv '{}'\nfrom: {}\nto: {}\ndata: {}"]"#,
            key,
            self.fqn,
            self.from
                .clone()
                .map(|actor| actor.to_string())
                .unwrap_or_default(),
            self.to
                .clone()
                .map(|actor| actor.to_string())
                .unwrap_or_default(),
            data
        )
    }
}

impl DrawDot for EventSend {
    fn draw(&self, key: EventKey) -> String {
        let data = serde_yaml::to_string(&self.payload).unwrap();
        format!(
            r#""{:?}" [label="send '{}'\nfrom: {}\nto: {}\ndata: {}"]"#,
            key,
            self.fqn,
            self.from,
            self.to
                .clone()
                .map(|actor| actor.to_string())
                .unwrap_or_default(),
            data
        )
    }
}

impl DrawDot for EventRespond {
    fn draw(&self, key: EventKey) -> String {
        format!(
            r#""{:?}" [label="respond '{}'\nfrom: {}"]"#,
            key,
            self.request_type,
            self.respond_from
                .clone()
                .map(|actor| actor.to_string())
                .unwrap_or_default(),
        )
    }
}
