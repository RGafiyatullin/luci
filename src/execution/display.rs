use std::fmt;

use slotmap::SlotMap;

use crate::execution::build::{BuildError, BuildErrorInner};
use crate::execution::runner::ReadyEventKey;
use crate::execution::sources::SingleScenarioSource;
use crate::execution::{Executable, KeyScenario, KeyScope, ScopeInfo, SourceCode};
use crate::recorder::{records as r, Record, RecordKind, RecordLog};
use crate::scenario::SrcMsg;

pub(super) struct DisplayRecord<'a> {
    pub(super) record:      &'a Record,
    pub(super) log:         &'a RecordLog,
    pub(super) executable:  &'a Executable,
    pub(super) source_code: &'a SourceCode,
}

impl<'a> fmt::Display for DisplayRecord<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let Self {
            record,
            log,
            executable,
            source_code,
        } = self;
        let (t0_wall, t0_rt) = log.t_zero;
        let (t_wall, t_rt) = record.at;
        let kind = &record.kind;

        let dt_wall = t_wall.duration_since(t0_wall);
        let dt_rt = t_rt.duration_since(t0_rt);
        write!(
            f,
            "[wall: {:>8?}; rt: {:>8?}] {}",
            dt_wall,
            dt_rt,
            DisplayRecordKind {
                kind,
                executable,
                source_code,
            }
        )
    }
}

impl<'a> fmt::Display for BuildError<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use BuildErrorInner::*;

        let Self {
            reason,
            scopes,
            sources,
        } = self;

        let scope = *match reason {
            UnknownEvent(_, k) => k,
            NotARequest(_, k) => k,
            UnknownActor(_, k) => k,
            UnknownDummy(_, k) => k,
            UnknownSubroutine(_, k) => k,
            UnknownFqn(_, k) => k,
            UnknownAlias(_, k) => k,
            DuplicateAlias(_, k) => k,
            DuplicateEventName(_, k) => k,
            DuplicateActorName(_, k) => k,
            DuplicateDummyName(_, k) => k,
        };

        write!(f, "{}", reason)?;
        fmt_scope_recursively(f, scope, scopes, sources)
    }
}

impl<'a> fmt::Debug for BuildError<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Display::fmt(self, f)
    }
}

pub(super) struct DisplayRecordKind<'a> {
    kind:        &'a RecordKind,
    executable:  &'a Executable,
    source_code: &'a SourceCode,
}

struct DisplayScope<'a> {
    scope:       KeyScope,
    executable:  &'a Executable,
    source_code: &'a SourceCode,
}

impl<'a> DisplayRecordKind<'a> {
    fn scope(&self, scope: KeyScope) -> DisplayScope<'a> {
        DisplayScope {
            scope,
            executable: self.executable,
            source_code: self.source_code,
        }
    }
}

impl<'a> fmt::Display for DisplayScope<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt_scope_recursively(
            f,
            self.scope,
            &self.executable.scopes,
            &self.source_code.sources,
        )
    }
}

impl<'a> fmt::Display for DisplayRecordKind<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use RecordKind::*;

        match self.kind {
            ProcessEventClass(r::ProcessEventClass(ReadyEventKey::Bind)) => {
                write!(f, "\x1b[90mrequested BIND\x1b[0m")
            },
            ProcessEventClass(r::ProcessEventClass(ReadyEventKey::RecvOrDelay)) => {
                write!(f, "\x1b[90mrequested RECV or DELAY\x1b[0m")
            },
            ProcessEventClass(r::ProcessEventClass(ReadyEventKey::Send(k))) => {
                let (scope, event) = self.executable.event_name((*k).into()).unwrap();
                write!(
                    f,
                    "\x1b[90mrequested SEND: {} ({})\x1b[0m",
                    event,
                    self.scope(scope)
                )
            },
            ProcessEventClass(r::ProcessEventClass(ReadyEventKey::Respond(k))) => {
                let (scope, event) = self.executable.event_name((*k).into()).unwrap();
                write!(
                    f,
                    "\x1b[90mrequested RESP: {} ({})\x1b[0m",
                    event,
                    self.scope(scope)
                )
            },

            ReadyBindKeys(r::ReadyBindKeys(ks)) => {
                write!(f, "\x1b[90mready binds: [")?;
                for k in ks {
                    let (scope, event) = self.executable.event_name((*k).into()).unwrap();
                    write!(f, " {}({}) ", event, self.scope(scope))?;
                }
                write!(f, "]\x1b[0m")
            },
            ReadyRecvKeys(r::ReadyRecvKeys(ks)) => {
                write!(f, "\x1b[90mready recvs: [")?;
                for k in ks {
                    let (scope, event) = self.executable.event_name((*k).into()).unwrap();
                    write!(f, " {}({}) ", event, self.scope(scope))?;
                }
                write!(f, "]\x1b[0m")
            },
            TimedOutRecvKey(r::TimedOutRecvKey(k)) => {
                let (scope, event) = self.executable.event_name((*k).into()).unwrap();
                write!(
                    f,
                    "\x1b[31mtimed out RECV: {} \x1b[0m({})",
                    event,
                    self.scope(scope)
                )
            },

            ProcessBindKey(r::ProcessBindKey(k)) => {
                let (scope, event) = self.executable.event_name((*k).into()).unwrap();
                write!(f, "process bind {} ({})", event, self.scope(scope))
            },
            ProcessSend(r::ProcessSend(k)) => write!(f, "process send {:?}", k),
            ProcessRespond(r::ProcessRespond(k)) => write!(f, "process resp {:?}", k),

            BindSrcScope(r::BindSrcScope(k)) => {
                write!(f, "\x1b[92msrc scope\x1b[0m {}", self.scope(*k))
            },
            BindDstScope(r::BindDstScope(k)) => {
                write!(f, "\x1b[92mdst scope\x1b[0m {}", self.scope(*k))
            },

            MatchActorAddress(r::MatchActorAddress(ka, ks, exp, act)) if exp == act => {
                let actor_name = &self.executable.actors[*ka].known_as[*ks];
                write!(
                    f,
                    "\x1b[32mMATCH ACTOR {} = {}\x1b[0m {}",
                    exp,
                    actor_name,
                    self.scope(*ks)
                )
            },
            MatchActorAddress(r::MatchActorAddress(ka, ks, exp, act)) => {
                let actor_name = &self.executable.actors[*ka].known_as[*ks];
                write!(
                    f,
                    "\x1b[33mMISMATCH ACTOR exp={}, act={}; {}\x1b[0m {}",
                    exp,
                    act,
                    actor_name,
                    self.scope(*ks)
                )
            },
            StoreActorAddress(r::StoreActorAddress(ka, ks, addr)) => {
                let actor_name = &self.executable.actors[*ka].known_as[*ks];
                write!(
                    f,
                    "\x1b[32mSET actor name {} = {} \x1b[0m {}",
                    addr,
                    actor_name,
                    self.scope(*ks)
                )
            },
            ResolveActorName(r::ResolveActorName(ka, ks, addr)) => {
                let actor_name = &self.executable.actors[*ka].known_as[*ks];
                write!(
                    f,
                    "resolve actor {} = {} {}",
                    addr,
                    actor_name,
                    self.scope(*ks)
                )
            },

            MatchDummyAddress(r::MatchDummyAddress(kd, ks, exp, act)) if exp == act => {
                let dummy_name = &self.executable.dummies[*kd].known_as[*ks];
                write!(
                    f,
                    "\x1b[32mMATCH DUMMY {} = {}\x1b[0m {}",
                    exp,
                    dummy_name,
                    self.scope(*ks)
                )
            },
            MatchDummyAddress(r::MatchDummyAddress(kd, ks, exp, act)) => {
                let dummy_name = &self.executable.dummies[*kd].known_as[*ks];
                write!(
                    f,
                    "\x1b[33mMISMATCH DUMMY exp={}, act={}; {}\x1b[0m {}",
                    exp,
                    act,
                    dummy_name,
                    self.scope(*ks)
                )
            },

            UsingMsg(r::UsingMsg(SrcMsg::Inject(name))) => write!(f, "msg.inj {:?}", name),
            UsingMsg(r::UsingMsg(SrcMsg::Literal(json))) => {
                write!(f, "msg.lit: {}", serde_json::to_string(&json).unwrap())
            },
            UsingMsg(r::UsingMsg(SrcMsg::Bind(bind))) => {
                write!(f, "msg.bind: {}", serde_json::to_string(&bind).unwrap())
            },

            BindToPattern(r::BindToPattern(pattern)) => {
                write!(f, "pattern: {}", serde_json::to_string(pattern).unwrap())
            },
            UsingValue(r::UsingValue(json)) => {
                write!(
                    f,
                    "\x1b[34mvalue: {}\x1b[0m",
                    serde_json::to_string(json).unwrap()
                )
            },
            NewBinding(r::NewBinding(key, value)) => {
                write!(
                    f,
                    "\x1b[32mSET {} = {}\x1b[0m",
                    key,
                    serde_json::to_string(value).unwrap()
                )
            },

            EventFired(r::EventFired(k)) => {
                let (scope, event) = self.executable.event_name(*k).unwrap();
                write!(
                    f,
                    "\x1b[1;32mcompleted {} \x1b[0m({})",
                    event,
                    self.scope(scope)
                )
            },

            SendMessageType(r::SendMessageType(fqn)) => {
                write!(f, "\x1b[36msend {}\x1b[0m", fqn)
            },
            SendTo(r::SendTo(None)) => write!(f, "\x1b[36mrouted\x1b[0m"),
            SendTo(r::SendTo(Some(addr))) => write!(f, "\x1b[36mto:{}\x1b[0m", addr),

            BindOutcome(r::BindOutcome(true)) => write!(f, "\x1b[1;32mBOUND\x1b[0m"),
            BindOutcome(r::BindOutcome(false)) => write!(f, "\x1b[33mNOT BOUND\x1b[0m"),

            EnvelopeReceived(r::EnvelopeReceived {
                message_name,
                from,
                to_opt,
            }) => {
                if let Some(to) = to_opt {
                    write!(
                        f,
                        "\x1b[35mreceived {} \x1b[1mfrom {} to {}\x1b[0m",
                        message_name, from, to
                    )
                } else {
                    write!(
                        f,
                        "\x1b[35mreceived {} \x1b[1mfrom {} routed\x1b[0m",
                        message_name, from
                    )
                }
            },

            MatchingRecv(r::MatchingRecv(k)) => {
                let (scope, event) = self.executable.event_name((*k).into()).unwrap();
                write!(f, "matching RECV: {} ({})", event, self.scope(scope))
            },

            ExpectedDirectedGotRouted(r::ExpectedDirectedGotRouted(name)) => {
                write!(f, "expected directed to {:?}, got routed", name)
            },

            Root => write!(f, "ROOT"),
            Error(r::Error { reason }) => write!(f, "{}", reason),
            // _fix_me => write!(f, "TODO"),
        }
    }
}

pub(super) fn fmt_scope_recursively(
    f: &mut fmt::Formatter<'_>,
    this_scope_key: KeyScope,
    scopes: &SlotMap<KeyScope, ScopeInfo>,
    sources: &SlotMap<KeyScenario, SingleScenarioSource>,
) -> fmt::Result {
    let this_scope = &scopes[this_scope_key];
    let this_source = &sources[this_scope.source_key].source_file;
    write!(f, "in {:?} ", &this_source)?;

    let mut invoked_as = this_scope.invoked_as.as_ref();
    while let Some((scope, event_name, _subroutine_name)) = invoked_as.take() {
        write!(f, "< {} ", event_name)?;
        invoked_as = scopes[*scope].invoked_as.as_ref();
    }
    Ok(())
}
