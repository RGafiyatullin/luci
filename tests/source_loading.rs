use insta::assert_debug_snapshot;
use luci::execution::SourceLoader;
use test_case::test_case;

#[test_case("00", "tests/source_loading/00-the-simplest-case.yaml", &["."])]
#[test_case("01", "tests/source_loading/01-one-inclusion.yaml", &["."])]
#[test_case("00.a", "00-the-simplest-case.yaml", &["tests/source_loading"])]
#[test_case("01.a", "01-one-inclusion.yaml", &["tests/source_loading"])]
#[test_case("02", "tests/source_loading/02-direct-cyclic-inclusion.yaml", &["."])]
#[test_case("02.a", "02-direct-cyclic-inclusion.yaml", &["tests/source_loading"])]
#[test_case("03", "03-indirect-cyclic-inclusion.yaml", &["tests/source_loading"])]
#[test_case("03", "04-diamond.yaml", &["tests/source_loading", "tests/source_loading/04-diamond"])]
fn load_sources(name: &str, main: &str, search_paths: &[&str]) {
    let mut loader = SourceLoader::new();
    loader.search_path = search_paths.into_iter().copied().map(From::from).collect();
    let outcome = loader.load(main);
    assert_debug_snapshot!(name, outcome);
}
