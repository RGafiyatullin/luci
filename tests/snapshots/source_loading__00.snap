---
source: tests/source_loading.rs
expression: outcome
---
Ok(
    (
        KeyScenario(
            1v1,
        ),
        {
            "./tests/source_loading/00-the-simplest-case.yaml": Source {
                source_file: "./tests/source_loading/00-the-simplest-case.yaml",
                subs: {},
                scenario: Scenario {
                    types: [],
                    subroutines: [],
                    cast: [],
                    events: [],
                    no_extra: NoExtra,
                },
            },
        },
    ),
)
